File: app/src/main/java/com/norvera/confession/ui/commandment/CommandmentsAdapter.java
package com.norvera.confession.ui.commandment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.norvera.confession.data.models.CommandmentEntry;
import com.norvera.confession.databinding.FragmentCommandmentsBinding;
import com.norvera.confession.interfaces.ClickListeners;

import androidx.annotation.NonNull;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

/**
 * {@link RecyclerView.Adapter} that can display a {@link CommandmentEntry} and makes a call to the
 * specified {@link ClickListeners.CommandmentClickListener}.
 */
public class CommandmentsAdapter extends ListAdapter<CommandmentEntry, CommandmentsAdapter.ViewHolder> {

    private static final DiffUtil.ItemCallback<CommandmentEntry> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<CommandmentEntry>() {
                @Override
                public boolean areItemsTheSame(@NonNull CommandmentEntry commandmentEntry, @NonNull CommandmentEntry t1) {
                    return (commandmentEntry.id == t1.id);
                }

                @Override
                public boolean areContentsTheSame(@NonNull CommandmentEntry commandmentEntry, @NonNull CommandmentEntry t1) {
                    return commandmentEntry.equals(t1);
                }
            };


    CommandmentsAdapter() {
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        FragmentCommandmentsBinding binding = FragmentCommandmentsBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);

        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder holder, int position) {
        CommandmentEntry commandmentEntry = getItem(position);
        holder.onBind(createOnClickListener(commandmentEntry.id), commandmentEntry);
    }


    private View.OnClickListener createOnClickListener(long commandmentId) {

        return view -> {
            NavDirections commandmentsFragmentDirections =
                    CommandmentsFragmentDirections.commandmentFragmentToExaminationFragment(commandmentId);
            Navigation.findNavController(view).navigate(commandmentsFragmentDirections);
        };

    }

    class ViewHolder extends RecyclerView.ViewHolder {

        FragmentCommandmentsBinding binding;

        ViewHolder(FragmentCommandmentsBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void onBind(View.OnClickListener clickListener, CommandmentEntry item) {
            itemView.setTag(item);
            binding.setClickListener(clickListener);
            binding.setCommandmentEntry(item);
            binding.executePendingBindings();
        }
    }

}

File: app/src/main/java/com/norvera/confession/ui/commandment/CommandmentsAdapter.kt
package com.norvera.confession.ui.commandment

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.Navigation
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.norvera.confession.data.models.CommandmentEntry
import com.norvera.confession.databinding.FragmentCommandmentsBinding

/**
 * [RecyclerView.Adapter] that can display a [CommandmentEntry] and makes a call to the
 */
internal class CommandmentsAdapter internal constructor() :
    ListAdapter<CommandmentEntry, CommandmentsAdapter.ViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = FragmentCommandmentsBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )

        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val commandmentEntry = getItem(position)
        holder.onBind(createOnClickListener(commandmentEntry.id), commandmentEntry)
    }


    private fun createOnClickListener(commandmentId: Long): View.OnClickListener {

        return View.OnClickListener {
            val commandmentsFragmentDirections =
                CommandmentsFragmentDirections.commandmentFragmentToExaminationFragment(
                    commandmentId
                )
            Navigation.findNavController(it).navigate(commandmentsFragmentDirections)
        }

    }

    internal inner class ViewHolder(var binding: FragmentCommandmentsBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun onBind(clickListener: View.OnClickListener, item: CommandmentEntry) {
            itemView.tag = item
            binding.clickListener = clickListener
            binding.commandmentEntry = item
            binding.executePendingBindings()
        }
    }

    companion object {

        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<CommandmentEntry>() {
            override fun areItemsTheSame(
                commandmentEntry: CommandmentEntry,
                t1: CommandmentEntry
            ): Boolean {
                return commandmentEntry.id == t1.id
            }

            override fun areContentsTheSame(
                commandmentEntry: CommandmentEntry,
                t1: CommandmentEntry
            ): Boolean {
                return commandmentEntry == t1
            }
        }
    }

}
